src/Components/Forms/src/EditContextDataAnnotationsExtensions.cs

```diff
@@ -16,98 +16,124 @@ namespace Microsoft.AspNetCore.Components.Forms
     {
-        private static ConcurrentDictionary<(Type ModelType, string FieldName), PropertyInfo?> _propertyInfoCache
-            = new ConcurrentDictionary<(Type, string), PropertyInfo?>();
-
         /// <summary>
         /// Adds DataAnnotations validation support to the <see cref="EditContext"/>.
         /// </summary>
         /// <param name="editContext">The <see cref="EditContext"/>.</param>
+        [Obsolete("Use " + nameof(EnableDataAnnotationsValidation) + " instead.")]
         public static EditContext AddDataAnnotationsValidation(this EditContext editContext)
         {
-            if (editContext == null)
-            {
-                throw new ArgumentNullException(nameof(editContext));
-            }
-
-            var messages = new ValidationMessageStore(editContext);
-
-            // Perform object-level validation on request
-            editContext.OnValidationRequested +=
-                (sender, eventArgs) => ValidateModel((EditContext)sender!, messages);
-
-            // Perform per-field validation on each field edit
-            editContext.OnFieldChanged +=
-                (sender, eventArgs) => ValidateField(editContext, messages, eventArgs.FieldIdentifier);
-
+            EnableDataAnnotationsValidation(editContext);
             return editContext;
         }
 
-        private static void ValidateModel(EditContext editContext, ValidationMessageStore messages)
+        /// <summary>
+        /// Enables DataAnnotations validation support for the <see cref="EditContext"/>.
+        /// </summary>
+        /// <param name="editContext">The <see cref="EditContext"/>.</param>
+        /// <returns>A disposable object whose disposal will remove DataAnnotations validation support from the <see cref="EditContext"/>.</returns>
+        public static IDisposable EnableDataAnnotationsValidation(this EditContext editContext)
+        {
+            return new DataAnnotationsEventSubscriptions(editContext);
+        }
+
+        class DataAnnotationsEventSubscriptions : IDisposable
         {
-            var validationContext = new ValidationContext(editContext.Model);
-            var validationResults = new List<ValidationResult>();
-            Validator.TryValidateObject(editContext.Model, validationContext, validationResults, true);
+            private static ConcurrentDictionary<(Type ModelType, string FieldName), PropertyInfo?> _propertyInfoCache
+                = new ConcurrentDictionary<(Type, string), PropertyInfo?>();
+
+            private readonly EditContext _editContext;
+            private readonly ValidationMessageStore _messages;
 
-            // Transfer results to the ValidationMessageStore
-            messages.Clear();
-            foreach (var validationResult in validationResults)
+            public DataAnnotationsEventSubscriptions(EditContext editContext)
             {
-                if (validationResult == null)
+                if (editContext == null)
```

Comment:
You could remove this since you're already checking for null as part of the assignment.

Reply:
Well spotted - fixed.